File: Data Structures & Algorithms/generate-parentheses/submission-0.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        stack = []

        def backtrack(op:int, cl:int) -> None:
            if cl == n and op == n:
                result.append("".join(stack))
                return
            
            # add a (
            if op < n:
                stack.append("(")
                backtrack(op+1, cl)
                stack.pop()
            if cl < op:
            # close )
                stack.append(")")
                backtrack(op,cl+1)
                stack.pop()
        
        backtrack(0,0)
        return result

File: Data Structures & Algorithms/simplify-path/submission-0.java
class Solution {
    public String simplifyPath(String path) {
        Stack<String> pathStack = new Stack<String>();
        int index = 0;
        String pathResult = "";

        path = path + "/";

        for (char c: path.toCharArray()){
            if (c == '/'){
                if (pathResult.equals("..")){
                    if (!pathStack.isEmpty()){
                        pathStack.pop();
                    }
                } else if (!pathResult.equals("") && !pathResult.equals(".")){
                    pathStack.push(pathResult);
                }
                pathResult = "";
            } else {
                pathResult = pathResult + c;
            }
        }


        return "/" + String.join("/", pathStack);
    }
}